listeners.js

```diff
@@ -199,32 +207,74 @@ module.exports = function startListeners(ssb, config, onError) {
       )
     )
 
-    // re-add missing people to a new epoch if the epoch creator didn't add everyone but they added us.
-    // we're only doing this for the preferred epoch atm
+    // recover from half-finished excludeMembers() calls
     pull(
       ssb.tribes2.list({ live: true }),
       pull.unique('id'),
       pull.map((group) =>
         pull(
           getPreferredEpoch.stream(group.id, { live: true }),
+          pull.unique('id'),
           pull.drain(
-            () => {
-              const timeoutScale = config.tribes2?.timeoutScale ?? 1000
-              const timeoutRandom = Math.random() * 25 + 5
-              const randomTimeout = timeoutScale * timeoutRandom
-
+            (preferredEpoch) => {
+              // re-add missing people to a new epoch if the epoch creator didn't add everyone but they added us.
+              // we're only doing this for the preferred epoch atm
+              const timeout = randomTimeout(config)
               const timeoutId = setTimeout(() => {
                 reAddMembers(ssb, group.id, null, (err) => {
                   // prettier-ignore
                   if (err && !isClosed) return onError(clarify(err, 'Failed re-adding members to epoch that missed some'))
                 })
-              }, randomTimeout)
-
+              }, timeout)
               closeCalls.push(() => clearTimeout(timeoutId))
+
+              // if we find an exclude and it's not excluding us but we don't find a new epoch, even after a while, then create a new epoch, since we assume that the excluder crashed or something
+              pull(
+                getMembers.stream(preferredEpoch.id, { live: true }),
+                pull.filter((members) => members.toExclude.length),
+                pull.take(1),
```

Comment:
I don't understand `{ live: true }` and `pull.take(1)` ... surely the live is pointless, as this stream gets killed after it first finds something? Maybe this is good in that it waits till it has at least one state?

You might be right, this just surprised/ confused me?

Reply:
also none of the data is used... you're just using this as a way to listen for a new membership event on the membership of an epoch?
I think this emits an initial state anyway right... so why not just run a setTimeout without all this?

Reply:
> surely the live is pointless

yeah i don't remember exactly but i think it might not have been emitting instantly so the live lets it wait some ms

Reply:
the live is very needed since the stream starts as soon as you find the current epoch, but we want to stream until we find an exclude, which might happen now/later/never

Reply:
i think a live db query for exclude messages for that epoch might be equivalent though. but can optimize that later

Reply:
> so why not just run a setTimeout without all this?

because we're specifically only interested in the event where an exclude was just published, for the preferred epoch

Reply:
i'd argue we can mark this conversation resolved?